x-pack/plugins/monitoring/server/plugin.ts

```diff
@@ -0,0 +1,401 @@
+    );
+    const cluster = (this.cluster = instantiateClient(
+      elasticsearchConfig,
+      this.log,
+      core.elasticsearch.createClient
+    ));
+
+    // Start our license service which will ensure
+    // the appropriate licenses are present
+    this.licenseService = new LicenseService().setup({
+      licensing: plugins.licensing,
+      monitoringClient: cluster,
+      config,
+      log: this.log,
+    });
+    await this.licenseService.refresh();
+
+    if (KIBANA_ALERTING_ENABLED) {
+      plugins.alerting.registerType(
+        getLicenseExpiration(
+          () => this.uiSettingsService,
+          cluster,
+          this.getLogger,
+          config.ui.ccs.enabled
+        )
+      );
+    }
+
+    // Register collector objects for stats to show up in the APIs
+    registerCollectors(
+      plugins.usageCollection,
+      config,
+      core.metrics.getOpsMetrics$(),
+      get(legacyConfig, 'kibana.index')
+    );
+
+    // If collection is enabled, create the bulk uploader
+    const kibanaMonitoringLog = this.getLogger(KIBANA_MONITORING_LOGGING_TAG);
+    const kibanaCollectionEnabled = config.kibana.collection.enabled;
+    if (kibanaCollectionEnabled) {
+      // Start kibana internal collection
+      const serverInfo = core.http.getServerInfo();
+      const bulkUploader = (this.bulkUploader = initBulkUploader({
+        elasticsearch: core.elasticsearch,
+        config,
+        log: kibanaMonitoringLog,
+        kibanaStats: {
+          uuid: core.uuid.getInstanceUuid(),
+          name: serverInfo.name,
+          index: get(legacyConfig, 'kibana.index'),
+          host: serverInfo.host,
+          transport_address: `${serverInfo.host}:${serverInfo.port}`,
+          port: serverInfo.port.toString(),
+          version: this.initializerContext.env.packageInfo.version,
+          snapshot: snapshotRegex.test(this.initializerContext.env.packageInfo.version),
+        },
+      }));
+
+      // Do not use `this.licenseService` as that looks at the monitoring cluster
+      // whereas we want to check the production cluster here
+      if (plugins.licensing) {
+        plugins.licensing.license$.subscribe((license: any) => {
+          // use updated xpack license info to start/stop bulk upload
+          const mainMonitoring = license.getFeature('monitoring');
+          const monitoringBulkEnabled =
+            mainMonitoring && mainMonitoring.isAvailable && mainMonitoring.isEnabled;
+          if (monitoringBulkEnabled) {
+            bulkUploader.start(plugins.usageCollection);
+          } else {
+            bulkUploader.handleNotEnabled();
+          }
+        });
+      }
+    } else {
+      kibanaMonitoringLog.info(
+        'Internal collection for Kibana monitoring is disabled per configuration.'
+      );
+    }
+
+    // If the UI is enabled, then we want to register it so it shows up
+    // and start any other UI-related setup tasks
+    if (config.ui.enabled) {
```

Comment:
Could you add TODO to move this to the `setup` phase? We want plugins to register all routes during `setup`.

Reply:
Done